routing/index_graph_serializer.hpp

```diff
@@ -0,0 +1,411 @@
+      for (Section & section : m_sections)
+        section.Deserialize(src);
+    }
+
+    uint32_t GetNumRoads() const { return m_numRoads; }
+    Joint::Id GetNumJoints() const { return m_numJoints; }
+    uint32_t GetNumSections() const { return m_sections.size(); }
+
+    Section const & GetSection(size_t index) const
+    {
+      CHECK_LESS(index, m_sections.size(), ());
+      return m_sections[index];
+    }
+
+    void AddSection(Section const & section) { m_sections.push_back(section); }
+
+  private:
+    uint8_t m_version = kVersion;
+    uint32_t m_numRoads = 0;
+    Joint::Id m_numJoints = 0;
+    vector<Section> m_sections;
+  };
+
+  class JointIdEncoder final
+  {
+  public:
+    template <class Sink>
+    void Write(Joint::Id originalId, BitWriter<Sink> & writer)
+    {
+      auto const & it = m_convertedIds.find(originalId);
+      if (it != m_convertedIds.end())
+      {
+        Joint::Id const convertedId = it->second;
+        if (convertedId >= m_count)
+        {
+          MYTHROW(RoutingException,
+                  ("Dublicated joint id should be less then count, convertedId =", convertedId,
+                   ", count =", m_count, ", originalId =", originalId));
+        }
+        writer.Write(kRepeatJointIdBit, 1);
+        // m_count - convertedId is less then convertedId in general.
+        // So write this delta to save file space.
+        WriteDelta(writer, m_count - convertedId);
+        return;
+      }
+
+      m_convertedIds[originalId] = m_count;
+      writer.Write(kNewJointIdBit, 1);
+      ++m_count;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    Joint::Id m_count = 0;
+    unordered_map<Joint::Id, Joint::Id> m_convertedIds;
+  };
+
+  class JointIdDecoder final
+  {
+  public:
+    // Joint::Id count is continuois along entire file.
+    // But deserializer skips some sections, therefore we should recover counter at each section
+    // start.
+    JointIdDecoder(Joint::Id startId) : m_count(startId) {}
+
+    template <class Source>
+    Joint::Id Read(BitReader<Source> & reader)
+    {
+      uint8_t const bit = reader.Read(1);
+      if (bit == kRepeatJointIdBit)
+      {
+        uint32_t const delta = ReadDelta(reader);
+        if (delta > m_count)
+          MYTHROW(RoutingException, ("Joint id delta", delta, "> count =", m_count));
+
+        return m_count - delta;
+      }
+
+      return m_count++;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    Joint::Id m_count;
+  };
+
+  // PointsFilter has two goals:
+  //
+  // 1. Deserialization skips some sections.
+  //    Therefore Joint::Id is not a continuous series of number [0, N).
+  //    PointsFilter converts joint ids to continuous series of number [0, N).
+  //
+  // 2. Some joints connects roads from different models.
+  //    E.g. 2 roads joint: all vehicles road + pedestrian road.
+  //    If we loads in car model, pedestrian road is skipped,
```

Comment:
If one loads car roads only, pedestrian roads mush be skipped, so joint would ...